Sources/FrontEnd/Parse/ParserState.swift

```diff
@@ -178,6 +178,30 @@ struct ParserState {
     }
   }
 
+  /// Adds up to `count` tokens from the lexer into the lookahead buffer and returns the number of added tokens.
+  @discardableResult private mutating func advanceLookahead(_ count: Int) -> Int {
+    for i in 0 ..< count {
+      guard let token = lexer.next() else { return i }
+      lookahead.append(token)
+    }
+
+    return count
+  }
+
+  /// Consumes and returns the first `kinds.count` tokens if they have the specified kinds.
+  mutating func take(_ kinds: Token.Kind...) -> [Token]? {
+    advanceLookahead(kinds.count)
+
+    if (lookahead.starts(with: kinds, by: {(token, kind) in token.kind == kind })) {
```

Comment:
I think this would be better expressed as:
```suggestion
    if peek(kinds.count).elementsEqual(kinds).by({ (a, b) in a.kind == b }) {
```
Then you don't need your `advanceLookahead` function.

It seems `ParserState.peek(_:)` was removed, but it should look like that:
```swift
mutating func peek(_ n: Int) -> Deque<Token>.SubSequence {
  while lookahead.count < n {
    guard let t = lexer.next() else { break }
    lookahead.append(t)
  }
  return lookahead.prefix(upTo: min(n, lookahead.count))
}
```